### PR TITLE
VMWare VM-Poweroff Experiment Enhancements

### LINKED ISSUE #447: VMWare VM-Poweroff Experiment Enhancements
## FEATURE REQUEST

VMWare VM-Poweroff experiment currently lacks the following functionalities:
1. The experiment lacks proper error handling during the API calls which can lead the experiment run to panic if something goes wrong during API calls.
2. The experiment lacks the functionality of injecting chaos in multiple VMs in serial or parallel mode.
3. The experiment lacks the logic for waiting through the duration for the VM to fully shut down or fully start before commencing the next step.

**Expected Functionality**
1. Proper error handling during VMWare API calls
2. Functionality of chaos injection in multiple VMs in serial and parallel mode
3. Logic for waiting through the duration for the VM to fully shut down or fully start before commencing the next step.

### PR DESCRIPTION
**Pull request by neelanjan00**
## What This PR Does
1. Adds proper error handling during the VMWare API calls to safely catch all the errors.
2. Adds functionality of injecting chaos in multiple VMs in serial or parallel mode.
3. Adds functionality logic for waiting through the duration for the VM to fully shut down or fully start before commencing the next step.
4. Updates serial chaos injection inline comments of VM and Disk chaos experiments of AWS, Azure, VMWare, and GCP.

## Which Issue This PR Fixes
fixes #447 

## Notes
**Positive Test Cases Covered**
1. Experiment run for single VM
2. Experiment run for multiple VMs
3. Experiment run for serial and parallel chaos injection

**Negative Test Cases Covered**
1. Experiment run for incorrect VcenterServer
2. Experiment run for incorrect VcenterUser and VcenterPass
3. Experiment run for wrong VM id
4. Experiment run for wrong Sequence

**Abort Test Cases Covered**
1. Experiment run abort before the chaos injection
2. Experiment run abort during the chaos injection
3. Experiment run abort after the chaos injection

### CODE REVIEW

**Comment by neelanjan00** [comment]
Experiment logs for parallel chaos injection for two VMs with Ids `vm-65` and `vm-85`:

```go
W0922 11:35:53.931316       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2021-09-22T11:35:53Z" level=info msg="Experiment Name: vm-poweroff"
time="2021-09-22T11:35:53Z" level=info msg="[PreReq]: Getting the ENV for the  experiment"
time="2021-09-22T11:35:53Z" level=info msg="[PreReq]: Updating the chaos result of vm-poweroff experiment (SOT)"
time="2021-09-22T11:35:57Z" level=info msg="[Info]: The Instance information is as follows" VM MOIDS="vm-65,vm-85" Ramp Time=0
time="2021-09-22T11:35:58Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (pre-chaos)"
time="2021-09-22T11:35:58Z" level=info msg="[Status]: No appLabels provided, skipping the application status checks"
time="2021-09-22T11:35:58Z" level=info msg="[Verification]: VMs are in running state (pre-chaos)"
time="2021-09-22T11:35:58Z" level=info msg="[Info]: Target VM Id list, [vm-65 vm-85]"
time="2021-09-22T11:35:58Z" level=info msg="[Chaos]: Stopping vm-65 VM"
time="2021-09-22T11:35:59Z" level=info msg="[Chaos]: Stopping vm-85 VM"
time="2021-09-22T11:36:00Z" level=info msg="[Wait]: Wait for VM 'vm-65' to get in POWERED_OFF state"
time="2021-09-22T11:36:00Z" level=info msg="[Status]: Checking vm-65 VM status"
time="2021-09-22T11:36:00Z" level=info msg="vm-65 VM state is POWERED_OFF"
time="2021-09-22T11:36:02Z" level=info msg="[Wait]: Wait for VM 'vm-85' to get in POWERED_OFF state"
time="2021-09-22T11:36:02Z" level=info msg="[Status]: Checking vm-85 VM status"
time="2021-09-22T11:36:02Z" level=info msg="vm-85 VM state is POWERED_OFF"
time="2021-09-22T11:36:04Z" level=info msg="[Wait]: Waiting for chaos interval of 30s"
time="2021-09-22T11:36:34Z" level=info msg="[Chaos]: Starting back vm-65 VM"
time="2021-09-22T11:36:34Z" level=info msg="[Chaos]: Starting back vm-85 VM"
time="2021-09-22T11:36:35Z" level=info msg="[Wait]: Wait for VM 'vm-65' to get in POWERED_ON state"
time="2021-09-22T11:36:35Z" level=info msg="[Status]: Checking vm-65 VM status"
time="2021-09-22T11:36:35Z" level=info msg="vm-65 VM state is POWERED_ON"
time="2021-09-22T11:36:37Z" level=info msg="[Wait]: Wait for VM 'vm-85' to get in POWERED_ON state"
time="2021-09-22T11:36:37Z" level=info msg="[Status]: Checking vm-85 VM status"
time="2021-09-22T11:36:37Z" level=info msg="vm-85 VM state is POWERED_ON"
time="2021-09-22T11:36:39Z" level=info msg="[Confirmation]: vm-poweroff chaos has been injected successfully"
time="2021-09-22T11:36:39Z" level=info msg="[Status]: Verify that the AUT (Application Under Test) is running (post-chaos)"
time="2021-09-22T11:36:39Z" level=info msg="[Status]: No appLabels provided, skipping the application status checks"
time="2021-09-22T11:36:39Z" level=info msg="[Status]: Verify that the IUT (Instance Under Test) is running (post-chaos)"
time="2021-09-22T11:36:39Z" level=info msg="[Verification]: VMs are in running state (post-chaos)"
time="2021-09-22T11:36:39Z" level=info msg="[The End]: Updating the chaos result of vm-poweroff experiment (EOT)"
```

File: chaoslib/litmus/gcp-vm-disk-loss/lib/gcp-vm-disk-loss.go [hunk 1]
```diff
@@ -137,6 +137,7 @@ func injectChaosInSerialMode(experimentsDetails *experimentTypes.ExperimentDetai
 			}
 
 			// run the probes during chaos
+			// the probes will be executed only once for all the target volumes
```
Comment (ispeakc0de) [inline chaoslib/litmus/gcp-vm-disk-loss/lib/gcp-vm-disk-loss.go:140]:
```suggestion
			// The OnChaos probes execution will start in the first iteration and keep running for the entire chaos duration 
```

File: experiments/vmware/vm-poweroff/experiment/vm-poweroff.go [hunk 2]
```diff
@@ -61,17 +61,17 @@ func VMPoweroff(clients clients.ClientSets) {
 	types.SetResultEventAttributes(&eventsDetails, types.AwaitedVerdict, msg, "Normal", &resultDetails)
 	events.GenerateEvents(&eventsDetails, clients, &chaosDetails, "ChaosResult")
 
-	//DISPLAY THE INSTANCE INFORMATION
+	//DISPLAY THE VM INFORMATION
 	log.InfoWithValues("[Info]: The Instance information is as follows", logrus.Fields{
-		"VM_INSTANCE_MOID": experimentsDetails.AppVMMoid,
-		"Ramp Time":        experimentsDetails.RampTime,
+		"VM MOIDS":  experimentsDetails.VMIds,
```
Comment (ispeakc0de) [inline experiments/vmware/vm-poweroff/experiment/vm-poweroff.go:66]: VM Instance IDs

Comment (neelanjan00) [reply experiments/vmware/vm-poweroff/experiment/vm-poweroff.go:66]: `VM MOIDS` is the correct key I think, which stands for "Virtual Machine Member Objects IDs"

File: chaoslib/litmus/gcp-vm-disk-loss/lib/gcp-vm-disk-loss.go (hunk 1 above)
Comment (neelanjan00) [reply chaoslib/litmus/gcp-vm-disk-loss/lib/gcp-vm-disk-loss.go]: Updated

**Review by ispeakc0de — APPROVED** [review]

File: chaoslib/litmus/vm-poweroff/lib/vm-poweroff.go [hunk 3]
```diff
@@ -37,70 +39,215 @@ func InjectVMPowerOffChaos(experimentsDetails *experimentTypes.ExperimentDetails
...
+			if experimentsDetails.EngineName != "" {
+				msg := "Injecting " + experimentsDetails.ExperimentName + " chaos in VM"
+				types.SetEngineEventAttributes(eventsDetails, types.ChaosInject, msg, "Normal", chaosDetails)
+				events.GenerateEvents(eventsDetails, clients, chaosDetails, "ChaosEngine")
+			}
+
+			for i, vmId := range vmIdList {
+
+				//Stopping the VM
+				log.Infof("[Chaos]: Stopping %s VM", vmId)
```
Comment (uditgaurav) [inline chaoslib/litmus/vm-poweroff/lib/vm-poweroff.go:94]: can we print vm name?

File: pkg/cloud/vmware/vm-operations.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+func WaitForVMStart(timeout, delay int, vcenterServer, vmId, cookie string) error {
+
+	log.Infof("[Status]: Checking %s VM status", vmId)
+	return retry.Times(uint(timeout / delay)).
+		Wait(time.Duration(delay) * time.Second).
+		Try(func(attempt uint) error {
+
+			vmStatus, err := GetVMStatus(vcenterServer, vmId, cookie)
+			if err != nil {
+				return errors.Errorf("failed to get %s VM status: %s", vmId, err.Error())
```
Comment (uditgaurav) [inline pkg/cloud/vmware/vm-operations.go:99]: can we print the name in place of moids for all such places?

**Review by uditgaurav — APPROVED** [review]

**Review by ispeakc0de — APPROVED** [review]

**Review by ksatchit — APPROVED** [review]

**Review by ispeakc0de — APPROVED** [review]

**Review by uditgaurav — APPROVED** [review]

**Pull request merged**